src/api/rest/decorators/with-reducers.js

```diff
@@ -0,0 +1,128 @@
+import * as Actions from 'api/rest/actions';
+import { normalize } from 'normalizr';
+import { Alert } from 'react-native';
+
+const displayError = error => {
+  Alert.alert('API Error', error);
+};
+
+export const withReducers = Provider => {
+  const client = new Provider();
+
+  return new Proxy(withReducers, {
```

Comment:
I'm sorry but I don't understand what is going on here =[ I think it's way too hard to follow, what about splitting functionality in small pieces that are easier to test/follow?

Reply:
@machour by the way, it's also hard for me to understand this yet, but I'll study it on the weekend. And why `withReducers`?

Reply:
@alejandronanez I'm currently extracting functions on my working copy :)

Here's some explanations:

The basic client, used like that:
`client.repos.getById("git-point")`
fetches() the url from github, and returns the Response (+ some meta data).
No redux, store, actions, anything involved. A simple Rest client.

`withReducers()` is a proxy around `client`, that serves the following purpose:

1. intercept the client call (`client.repos.getById("git-point")`)
2. compute the corresponding action name/pagination name automatically (reduce boiler plate, standardize naming)
3. check if the requested data is already in store, if that's the case, return; 
4. if the data is not available:
  - If it's a paginated call, check the pagination status, and optionally override the API URL that will be called
5. dispatch PENDING & trigger the real request (as it would have been the case without `withReducers()`
6. normalize the received JSON data, and trigger SUCCESS for it to be stored

It's a lot of things indeed, but it's actually the WHOLE stuff, centralized in one place.

Don't hesitate if you need clarifications mate (/cc @lex111)